File: examples/simple_pub_sub.rs
//! Simple publisher-subscriber example for MiniROS
//! 
//! This example demonstrates basic pub/sub functionality:
//! - Creates a publisher node that sends string messages
//! - Creates a subscriber node that receives and prints messages
//! - Shows how to use async callbacks for message processing

use mini_ros::prelude::*;
use mini_ros::message::StringMsg;
use std::time::Duration;
use tokio::time::sleep;
use tracing::{info, warn};

#[tokio::main]
async fn main() -> Result<()> {
    // Initialize tracing for logging
    tracing_subscriber::fmt::init();
    
    info!("Starting MiniROS simple pub/sub example");

    // Create publisher node
    let mut pub_node = Node::new("publisher_node")?;
    pub_node.init().await?;
    
    // Create subscriber node  
    let mut sub_node = Node::new("subscriber_node")?;
    sub_node.init().await?;

    // Create publisher and subscriber
    let publisher = pub_node.create_publisher::<StringMsg>("chatter").await?;
    let subscriber = sub_node.create_subscriber::<StringMsg>("chatter").await?;

    // Set up message callback
    subscriber.on_message(|msg: StringMsg| {
        info!("Received message: {}", msg.data);
    })?;

    // Give some time for discovery and connection setup
    sleep(Duration::from_secs(1)).await;

    // Publish some messages
    info!("Publishing messages...");
    for i in 0..10 {
        let message = StringMsg {
            data: format!("Hello, MiniROS! Message #{}", i),
        };
        
        match publisher.publish(&message).await {
            Ok(_) => info!("Published: {}", message.data),
            Err(e) => warn!("Failed to publish message: {}", e),
        }
        
        sleep(Duration::from_millis(500)).await;
    }

    // Wait a bit for final messages to be processed
    sleep(Duration::from_secs(2)).await;

    // Shutdown nodes
    info!("Shutting down...");
    pub_node.shutdown().await?;
    sub_node.shutdown().await?;

    info!("Example completed successfully");
    Ok(())
} 

File: python/examples/simple_pubsub.py
#!/usr/bin/env python3
"""
Simple Pub/Sub Demo for miniROS-rs

This single-file example demonstrates both publishing and subscribing
in the most minimal way possible - under 30 lines of code!
Perfect example of the "mini" philosophy.
"""

import mini_ros
import time
import threading

def main():
    # Initialize miniROS
    mini_ros.init()
    
    # Create node
    node = mini_ros.Node('simple_demo')
    
    # Create publisher and subscriber for same topic
    pub = node.create_publisher(mini_ros.StringMessage, 'hello', 10)
    
    def on_message(msg):
        print(f'📨 Received: {msg.data}')
    
    sub = node.create_subscription(mini_ros.StringMessage, 'hello', on_message, 10)
    
    print('🚀 Starting miniROS pub/sub demo...')
    
    # Publish some messages
    for i in range(3):
        msg = mini_ros.StringMessage()
        msg.data = f'Hello miniROS #{i+1}'
        pub.publish(msg)
        print(f'📤 Published: {msg.data}')
        time.sleep(0.5)
    
    print('✅ Demo completed!')
    mini_ros.shutdown()

if __name__ == '__main__':
    main() 